Full Statistical Synchronization

Baldie-dev opened the pull request:
This is part of the #92 .
Do not hesitate to ask about anything, I made kind a lot of changes.

High-level overview:

- When a client opens the Statistical window, it will send sync packet
- Host will send full statistical data, plus it will subscribe a player for regular updates.
- Every game-tick, the host will create a snapshot of changes in statistics in that game-tick.
- Every second, the host will sent regular updates (array of snapshots) to subscribers (players with stats window opened).
- When a player closes the window, it will send an "unsubscribe" packet.

Video on [youtube](https://www.youtube.com/watch?v=77XVlXyIDWY).
Bandwidth is around 7KB/s for the late-game savegame with 7 active factories.

Changes to the existing code:

- `GameData_Patch.cs` I needed to have synchronized the indexes in array for factories between client and server (Host's GameData.factories[i] = Clients GameData.factories[i]). Since this index is used in statistics to determine which factory belongs to which planet, also these indexes are in variables such as (Factory.factoryIndex, PlanetData.factoryIndex, ...) and they should be the same between client and host.
  - So the change is basically, that host will send an extra `int` value that will say how many factories exist (to populate `GameData.factoryCount`) and a second `int` value that will indicate an array index on which the requested factory has to be stored.

hubastard commented on NebulaPatcher/Patches/Transpilers/UIProductionStatWindow_Patch.cs:
```diff
@@ -0,0 +1,140 @@
+                    codes[i + 3] = new CodeInstruction(OpCodes.Call, AccessTools.Method(typeof(StatisticsManager), "UpdateTotalChargedEnergy"));
+                }   
+            }
+            //return codes.AsEnumerable();
```
remove old commented code

Baldie-dev replied:
fixed :)

hubastard commented on NebulaPatcher/Patches/Transpilers/UIProductionStatWindow_Patch.cs:
```diff
@@ -0,0 +1,140 @@
+
+        public static IEnumerable<CodeInstruction> ReplaceFactoryCondition(IEnumerable<CodeInstruction> instructions)
+        {
+            //change: if (starData.planets[j].factory != null)
+            //to    : if (starData.planets[j].factoryIndex != -1)
```
I really like the way you formatted this comment, could you add more comments like this to the changes that `UpdatePower_Transpiler` does under the hood. It would really help with maintainability.

Baldie-dev replied:
I have added a comment there about it.

Baldie-dev commented on NebulaModel/DataStructures/StatisticalSnapshot.cs:
```diff
@@ -0,0 +1,90 @@
+    {
+        //List of statistical changes for each planet that happend in one gameTick
+        public List<ProductionChangeStruct>[] ProductionChangesPerFactory;
+        public long[] PowerGenRegister;
+        public long[] PowerConRegister;
+        public long[] PowerChaRegister;
+        public long[] PowerDisRegister;
```
To make it less confusing, I have used the same names that are used in the original code, so it would be obvious at which parts it is being used. But I can change to to full names: "Generation", "Consumption","Charging","Discharging".

![image](https://user-images.githubusercontent.com/79051050/113354987-5e0c8180-9340-11eb-96b6-c785263df786.png)

hubastard replied:
Ok yeah, I think I still prefer to have the complete word in our internal data structures

Baldie-dev replied:
fixed :)

Baldie-dev commented on NebulaHost/MultiplayerHostSession.cs:
```diff
@@ -70,11 +74,18 @@ public void SendPacket<T>(T packet) where T : class, new()
             PlayerManager.SendPacketToAllPlayers(packet);
         }
+        
+        private void FixedUpdate()
+        {
+            StatisticsManager.CaptureStatisticalSnapshot();
+        }
```
oh, I agree, I will move it there.

Baldie-dev replied:
fixed :)

Baldie-dev commented on NebulaClient/PacketProcessors/Statistics/StatisticsDataProcessor.cs:
```diff
@@ -0,0 +1,28 @@
+    {
+        public void ProcessPacket(StatisticsDataPacket packet, NebulaConnection conn)
+        {
+            Log.Info($"Parsing Statistics data from the server.");
```
I think removing it from the code would be better. Since there is not much information given in the logs by these statements.

Baldie-dev replied:
fixed :)

Baldie-dev commented:
Following commits were made:

- Merging with current PR and changing current BinaryWriter / Reader.
- Resolving merge conflicts.
- Addressing PR comments.

I have just tested it and everything seems to be working.
